testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -0,0 +1,538 @@
+          @Override
+          public void run() {
+            try {
+              Container c = container;
+              if (c != null) {
+                c.stop();
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception during global container shutdown: {}", e.getMessage(), e);
+            }
+          }
+        };
+
+    Runtime.getRuntime().addShutdownHook(containerShutdown);
+  }
+
+  private static Container container() {
+    if (container == null) {
+      throw new IllegalStateException("Stargate container has not been configured");
+    }
+
+    return container;
+  }
+
+  private static void enter(AtomicBoolean flag) {
+    if (!flag.compareAndSet(false, true)) {
+      throw new IllegalStateException(
+          String.format(
+              "Concurrent execution with %s is not supported",
+              StargateContainer.class.getSimpleName()));
+    }
+  }
+
+  private static void leave(AtomicBoolean flag) {
+    flag.compareAndSet(true, false);
+  }
+
+  private static StargateParameters parameters(StargateSpec spec, ExtensionContext context)
+      throws Exception {
+    Builder builder = StargateParameters.builder();
+
+    String customizer = spec.parametersCustomizer().trim();
+    if (!customizer.isEmpty()) {
+      Object testInstance = context.getTestInstance().orElse(null);
+      Class<?> testClass = context.getRequiredTestClass();
+      Method method = testClass.getMethod(customizer, Builder.class);
+      method.invoke(testInstance, builder);
+    }
+
+    return builder.build();
+  }
+
+  private static synchronized void start(ExtensionContext context) throws Exception {
+    ClusterConnectionInfo backend = ExternalStorage.connectionInfo();
+
+    StargateSpec spec =
+        context
+            .getElement()
+            .flatMap(e -> Optional.ofNullable(e.getAnnotation(StargateSpec.class)))
+            .orElseGet(
+                () ->
+                    context
+                        .getTestClass()
+                        .flatMap(c -> Optional.ofNullable(c.getAnnotation(StargateSpec.class)))
+                        .orElse(DefaultSpecHolder.get()));
+
+    StargateParameters params = parameters(spec, context);
+
+    Container c = container;
+    if (c != null) {
+      if (!c.matches(backend, spec, params)) {
+        c.stop();
+      } else {
+        // reuse container
+        LOG.info("Reusing Stargate container for the next test.");
+        return;
+      }
+    }
+
+    c = new Container(backend, spec, params);
+    c.start();
+    container = c;
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    enter(suiteExecuting);
+    start(context);
+  }
+
+  @Override
+  public void afterAll(ExtensionContext context) {
+    leave(suiteExecuting);
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) throws Exception {
+    enter(testExecuting);
+    start(context);
```

Comment:
The first test method we hit will restart an instance if its config doesn't match the test class's. That's a bit wasteful when you consider this example:
```java
@UseStargateContainer public class ExtensionTest {
  @Test @StargateSpec(nodes = 2) public void doSomething1() {}
  @Test @StargateSpec(nodes = 1) public void doSomething2() {}
}
```

The extension will perform the following operations:
```
>>> Enter start for Optional[class io.stargate.it.ExtensionTest], 1 nodes needed
>>> Start new Stargate
>>> Enter start for Optional[public void io.stargate.it.ExtensionTest.doSomething1()], 2 nodes needed
>>> Stop current Stargate because it doesn't match config (1 nodes)
>>> Start new Stargate
>>> Enter start for Optional[public void io.stargate.it.ExtensionTest.doSomething2()], 1 nodes needed
>>> Stop current Stargate because it doesn't match config (2 nodes)
>>> Start new Stargate
```

Note that we start an instance at the beginning of the class, but it's immediately shut down as we enter the first method.

One way to fix this is to not start anything in `beforeAll()`. Instead, just memorize the class's `StargateSpec`, and use it as the fallback in the next `beforeEach` if the test method doesn't declare its own. Also with this approach you would only need one AtomicBoolean for mutual exclusion.
The only downside is that if the test class declares its own `@BeforeAll` method, the Stargate instance won't be started when it executes, which might be a bit counter-intuitive.

Another way is to let the caller decide how they want to use the extension: we'd only do "per method" if the extension is declared as an [instance field](https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic-instance-fields); we can detect that in the extension because beforeAll/afterAll will not be called. I think it would be more correct, but a bit more complicated.

Reply:
I'm thinking to just drop the default `@StargateSpec` completely. If a test needs a SG node in `@BeforeAll` it will have to have an annotation at the class level. If the annotation is missing, the extension will simply not start SG. Consequently, if SG connection info parameters are required, but there is not spec in scope, we can produce a nice error message.

Also, I'll probably drop the boolean mutex check and replace it with a check against the cluster ID :thinking:

Reply:
Right, that would work too. The only drawback is that `@StargateSpec` would now have to be explicitly set everywhere.

Reply:
What if we introduced a `scope` for `@StargateSpec` (similar to `@ClusterSpec`), and added a new scope of `VOID`. This way if methods had a variety of specs the class could have the `VOID` spec to effectively stop any previous SG containers without starting new ones. WDYT?

Reply:
I don't know, I feel that JUnit already has a pretty clear scoping model to execute extensions per class or per method, so why add our own on top.

Reply:
Do you mean that we derive scope from the element that has `@StargateSpec`? So, e.g. if the annotation is on the class, we start/stop in before/afterAll callbacks, if it's on the method we start/stop in before/afterEach callbacks? Concequently, if it's default (no `@StargateSpec` annotation) - it's a shared instance stopped at the JVM exit?

Reply:
start/stop logic has been reworked. Please review again.